### PR TITLE
feat: ION DID Provider implementation

### LINKED ISSUE #336: Add support for did:ion
Add support for did:ion

### PR DESCRIPTION
**Pull request by nklomp**
## What issue is this PR fixing

closes #336 and closes #440

## What is being changed
Implement ION DID Provider. Seeking initial feedback before rebasing to Veramo and opening a PR there.

Changes:

- Do not depend on ion-tools
- Depend on external ION POW SDK. 
- Update and recovery keys, including key rotations
- Full implementation of all Identifier methods, with the exception of the recent updateIdentifier method

Known issues:
- ~~The ION PoW lib on which this PR depends works on Node. It is not tested yet on React-native, but we do have a React-Native version of the dependency. I will ensure it will work on both, since we need it @Sphereon anyway~~
- Adding/Updating services/keys does not get propagated on chain. Needs more investigation. Microsoft gives back a 200 response for them, but it never ends up in the DID Document.

## Quality
Check all that apply:
* [X] I want these changes to be integrated
* [X] I successfully ran `yarn`, `yarn build`, `yarn test`, `yarn test:browser` locally.
* [X] I allow my PR to be updated by the reviewers (to speed up the review process).
* [X] I added unit tests.
* [X] I added integration tests.
* [ ] I did not add automated tests because _________, and I am aware that a PR without tests will likely get rejected.

## Details
If applicable, add screen captures, error messages or stack traces to help explain your problem.

### CODE REVIEW

**Comment by nklomp** [comment]
Updated our ION POW package to now depend on a new isomorphic Argon2 hash package created by us. Verified to be working in browser, node and react-native. For react-native a peer dependency is put in the package.json. This dep needs to be installed, since native components are needed for Argon2 in RN. Except for installing the peer dep, which is mentioned in the Readme, a user does not have to change anything else on RN.

Still looking into the update issue. As far as we can tell right now this only is happening on the public Microsoft hosted ION node. I will get in contact with my contacts within the Microsoft team to see if this is a known issue or not. In the mean time we are setting up our own ION node, but that takes a bit of time, given the components involved.

File: __tests__/shared/didCommWithEthrDidFlow.ts [hunk 1]
```diff
@@ -134,6 +134,106 @@ export default (testContext: {
       expect([...(resolution?.didDocument?.service || [])]).toEqual([])
     })
 
+    it('should add dummy service 2 to identifier', async () => {
```
Comment (mirceanis) [inline __tests__/shared/didCommWithEthrDidFlow.ts:137]: @nklomp I'm not sure why these changes are showing up in your PR.
Is it possible to rebase on top of our latest `next` branch, or merge it into your branch?

Comment (nklomp) [reply __tests__/shared/didCommWithEthrDidFlow.ts:137]: Hmzz, odd, If i remember correctly it was rebased on @next. Will merge next in to fix

Comment (mirceanis) [reply __tests__/shared/didCommWithEthrDidFlow.ts]: thank you!

File: packages/did-provider-ion/src/functions.ts [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+ * @param ionMeta
+ */
+export const tempMemoryKey = async (
+  type: KeyType.Ed25519 | KeyType.Secp256k1 | KeyType,
+  privateKeyHex: string,
+  kid: string,
+  kms: string,
+  ionMeta: IonKeyMetadata
+): Promise<IKey> => {
+  const tmpKey = (await new KeyManagementSystem(new MemoryPrivateKeyStore()).importKey({
```
Comment (mirceanis) [inline packages/did-provider-ion/src/functions.ts:220]: This is a good candidate for a utility function. It would then allow us to decouple this did-provider from the kms-local implementation.

**Review by mirceanis — APPROVED** [review]
Thank you for the contribution!

There is a tiny nit-pick about the import paths in one of the tests, but otherwise I wish all PRs were this beautiful :)

File: packages/did-provider-ion/src/functions.ts (hunk 2 above)
Comment (nklomp) [reply packages/did-provider-ion/src/functions.ts:220]: Add a ticket for it https://github.com/uport-project/veramo/issues/1021

**Comment by nklomp** [comment]
> Thank you for the contribution!
> 
> There is a tiny nit-pick about the import paths in one of the tests, but otherwise I wish all PRs were this beautiful :)
Thanks. Updated the PR, and created tickets for the TD: https://github.com/uport-project/veramo/issues/1019  and suggestion: https://github.com/uport-project/veramo/issues/1021

**Comment by nklomp** [comment]
Updated Veramo dep versions as it gave an error on the latest CI run

**Comment by mirceanis** [comment]
> Updated Veramo dep versions as it gave an error on the latest CI run

Thanks, I was trying to do that on my end but I didn't know if I could push to your PR branch.
If this builds fine, we'll be merging it soon
